paddle/framework/tensor.cc

```diff
@@ -0,0 +1,95 @@
+#include "paddle/framework/tensor.h"
+
+namespace paddle {
+namespace framework {
+
+template <typename T>
+const T* Tensor::Data() const {
+  PADDLE_ASSERT(holder_ != nullptr);
+  PADDLE_ASSERT(holder_->Place() == place_);
+  PADDLE_ASSERT(holder_->Size() >= product(dims_) * sizeof(T));
+  return static_cast<const T*>(holder->Ptr());
+}
+
+bool Tensor::NeedReset() {
+  return (holder_ == nullptr || holder_->Place() != place_ ||
+          holder_->Size() < product(dims_) * sizeof(T));
+}
+
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+T* Tensor::MutableData() {
+  if (NeedReset()) {
+    holder_.reset(new PlaceholderImpl(place_, product(dims_) * sizeof(T)));
+  }
+  return static_cast<T*>(holder_->Ptr());
+}
+
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+T* Tensor::MutableData(const DDim& dims) {
+  dims_ = dims;
+  return MutableData<T>();
+}
+
+template <typename T, typename = std::enable_if<std::is_pod<T>::value>::type>
+T* Tensor::MutableData(const DDim& dims, const Place& place) {
+  dims_ = dims;
+  place_ = place;
+  return MutableData<T>();
+}
+
+int Tensor::Rank() const { return arity(dims_); }
+
+int Tensor::Numel() const { return product(dims_); }
+
+void Tensor::Resize(const DDim& dims) {
+  dims_ = dims;
+  return;
+}
+
+void Tensor::Reshape(const DDim& dims) {
+  if (product(dims) != product(dims_)) {
+    // TODO: error: "Reshape() can not change tensor's numel".
+  }
+  _dims = dim;
+  return;
+}
+
+void Tensor::ShareData(const Tensor& src) {
+  if (src.NeedReset()) {
+    // TODO: error: "Src tensor need to be reseted before calling ShareData()".
+  }
+  holder_ = src.Holder();
+  dims_ = src.Dims();
+  place_ = src.Place();
+  return;
+}
+
+template <typename T>
+void CopyFrom(const Tensor& src) {
+  if ((void*)&src == (void*)this) {
```

Comment:
There are a few checking in `Tensor::Data()`. I can't find out more potential problems for the moment... Do you have any idea?